core/textile/model/buckets.go

```diff
@@ -0,0 +1,158 @@
+package model
+
+import (
+	"context"
+	"errors"
+
+	"github.com/FleekHQ/space-daemon/log"
+	"github.com/textileio/go-threads/api/client"
+	core "github.com/textileio/go-threads/core/db"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/go-threads/db"
+	"github.com/textileio/go-threads/util"
+)
+
+type BucketSchema struct {
+	ID     core.InstanceID `json:"_id"`
+	Slug   string          `json:"slug"`
+	Backup bool            `json:"backup"`
+	DbID   string
+}
+
+const bucketModelName = "BucketMetadata"
+
+var errBucketNotFound = errors.New("Bucket not found")
+
+func (m *model) CreateBucket(ctx context.Context, bucketSlug, dbID string) (*BucketSchema, error) {
+	log.Debug("Model.CreateBucket: Storing bucket " + bucketSlug)
+	if existingBucket, err := m.FindBucket(ctx, bucketSlug); err == nil {
+		log.Debug("Model.CreateBucket: Bucket already in collection")
+		return existingBucket, nil
+	}
+
+	log.Debug("Model.CreateBucket: Initializing db")
+	metaCtx, metaDbID, err := m.initBucketModel(ctx)
+	if err != nil && metaDbID == nil {
+		return nil, err
+	}
+
+	newInstance := &BucketSchema{
+		Slug:   bucketSlug,
+		ID:     "",
+		DbID:   dbID,
+		Backup: true,
+	}
+
+	instances := client.Instances{newInstance}
+	log.Debug("Model.CreateBucket: Creating instance")
+
+	res, err := m.threads.Create(metaCtx, *metaDbID, bucketModelName, instances)
+	if err != nil {
+		return nil, err
+	}
+	log.Debug("Model.CreateBucket: stored bucket with dbid " + newInstance.DbID)
+
+	id := res[0]
+	return &BucketSchema{
+		Slug:   newInstance.Slug,
+		ID:     core.InstanceID(id),
+		DbID:   newInstance.DbID,
+		Backup: newInstance.Backup,
+	}, nil
+}
+
+func (m *model) UpsertBucket(ctx context.Context, bucketSlug, dbID string) (*BucketSchema, error) {
```

Comment:
Since there are more fields coming on this schema from the mirror bucket work, maybe we make this take in a bucket schema object so you can set any field.  cc @maurycy

Reply:
@jsonsivar 

Yes, I love the `model` separation. We even discussed this today, ie, moving different collections into different files in the `collection` package.

When it comes to the `func` signatures, I have mixed feelings. I have created a separate `struct` for an upsert myself and I don't like exposing, ie, returning, the internal `*Schema` (the one with `json`), but not really sure if we should multiplicate the `struct` for every CRUD.